### PR TITLE
F54 vote once only

### PR DESCRIPTION
**Pull request by gmcgibbon**
This adds a Vote table to track individual votes on ideas. It also uses validations for a user to only be able to vote once per idea.

### CODE REVIEW

File: app/controllers/ideas_controller.rb [hunk 1]
```diff
@@ -21,9 +21,13 @@ def create
...
 
   def upvote
-    @idea = Idea.find(params[:id])
-    @idea.upvote!
-    render json: {votes: @idea.votes}
+    begin
+      @idea = Idea.find(params[:id])
+      @idea.upvote current_user
+      render json: {total_votes: @idea.total_votes}
+    rescue
```
Comment (swalberg) [inline app/controllers/ideas_controller.rb]: Can you be more specific about what you're `rescue`ing? If you grab everything, we might miss some legitimate problems.

File: db/schema.rb [hunk 2]
```diff
@@ -11,18 +11,19 @@
 #
 # It's strongly recommended to check this file into your version control system.
 
-ActiveRecord::Schema.define(:version => 20140313234621) do
+ActiveRecord::Schema.define(:version => 20140427222532) do
```
Comment (swalberg) [inline db/schema.rb]: Thank you for indexing! :dancers:

**Comment by JamesPartaker** [comment]
Here are a few things I seen when looking through the code
1. I'm not sure if you should be using an exception in the ideas controller to handle the case where the user has already voted on the idea. It seems a little weird to me and I think it would be clearer to call vote.save instead of vote.save! in the idea model's upvote method (which returns false instead of throwing an exception). Then in the the upvote action just have an if statement to check if the upvote succeeded.
2. I'm not sure why you added vote_id fields to the ideas and users tables. They are not being used anywhere.
3. Try adding a counter cache to the ideas table on the number of votes. While it's not necessary it could make things quicker.
4. It is good that you validated the uniqueness of the user for each of the votes inside the vote model but you can also enforce this within the database as well. I think that there are situations where validating the uniqueness only within the rails app can lead to race conditions. You can do this by adding an index on both the idea_id and the user_id then adding a unique constraint on it.

```
add_index(:votes, [:idea_id, :user_id], unique: true)
```

Also you could add, null: false to make sure they idea_id and user_id aren't empty.

I would be interested in hearing what other people think.

File: app/models/idea.rb [hunk 3]
```diff
@@ -4,14 +4,19 @@ class Idea < ActiveRecord::Base
   validates_presence_of :description, :title
 
   belongs_to :project
+  has_many :votes
 
-  def upvote!
-    self.votes += 1
-    save
+  def upvote user
```
Comment (swalberg) [inline app/models/idea.rb]: style: `def upvote(user)`

File: app/models/idea.rb [hunk 4]
```diff
@@ -4,14 +4,19 @@ class Idea < ActiveRecord::Base
...
+    vote = self.votes.build(user_id: user.id)
+    vote.save!
```
Comment (swalberg) [inline app/models/idea.rb:11]: Yea, see @JamesPartaker 's note below, that's probably the more elegant way to handle it.

File: app/controllers/ideas_controller.rb [hunk 5]
```diff
@@ -21,9 +21,13 @@ def create
   end
 
   def upvote
-    @idea = Idea.find(params[:id])
-    @idea.upvote!
-    render json: {votes: @idea.votes}
+    begin
+      @idea = Idea.find(params[:id])
+      @idea.upvote current_user
```
Comment (swalberg) [inline app/controllers/ideas_controller.rb]:
```
if @idea.upvote current_user
  render json: "woohoo"
else
  render json: "doh"
end
```

**Comment by amirci** [comment]
My comments:
- _vote_ vs _vote!_: I think that both are valid and both are idiomatic to RoR. I think that without the _if_ looks good.

``` ruby
def upvote
    @idea = find .....
    @idea.upvote!
    render :json, {votes: .... }
rescue
   render :json, 'some error'
end
```
- Cache: I would leave it for another feature to do the change a making sure we need it
- What would be the situation that will cause a race condition?

**Comment by JamesPartaker** [comment]
Here is a blog post that talks about the race conditions on uniqueness validations. Generally I think it's just a good idea to maintain the data integrity within the database. Which is also why I recommended adding null: false to the foreign keys as well.

http://robots.thoughtbot.com/the-perils-of-uniqueness-validations

**Comment by gmcgibbon** [comment]
Thanks for all the feedback guys!

Not entirely sure why it shows the build as failing for Travis CI... It was all passing with guard on my last commit. 

My rule of thumb for method! vs method is: if the method changes the invoked object's state in any way, it is dangerous and needs ! at the end. Similar to if a method returns a bool it should have ? at the end. Idea's upvote method no longer changes state of the idea so I removed the ! 
Since I am using save! does the upvote method need ! to indicate it raises exceptions?

Which style should I use for the upvote action? An if block or an exception rescue? 

I see the unused columns and think they are leftovers from some tests with the schema I did before, sorry about that. They will be removed asap. I have never done database level validations with ActiveRecord, can someone explain why the unique validation needs to be done when adding the associated index? I read that database indexing is done to enable binary searching to increase query performance.

Race conditions are an interesting thought but I'm not sure what would cause them in this scenario. 
- A user is logged into their account and tries to upvote. 
- An Ajax call is sent, processed, and the response updates the count.
- The new entry in Vote is created and rejects new votes on the same project from the same user
  There is no competition because the validation only raises on votes from a user after they have sent their first vote request. I may have this wrong but I think that makes logical sense.

**Comment by JamesPartaker** [comment]
You can only add the uniqueness constraint with an index. My best guess is that this is because otherwise, you would have to do a linear search through the entire table on every insert to make sure that the value doesn't already exist. This would be a slow operation on a table with millions+ of records. 

The race condition exists between the time that rails checks to see if a vote exists (an SQL SELECT) and when it inserts it. If two app servers ask the database "at the same time" if a vote exists then they will both get back "nope, you're good" and they will both happily go on their way making the insert. But, now you have two votes with both the same idea and user.

This situation is very improbable and in this case isn't that big of deal in terms of what could go wrong if this does occur. But there are other situations were you would definitely want to make sure that this doesn't happen.

**Comment by amirci** [comment]
I would say that if you are using _save!_ then add do _upvote!_ to indicated you are going to throw an exception.

For the action I would prefer to avoid the _if_ and use the _rescue_. Nesting is harder to read.

**Comment by gmcgibbon** [comment]
Sorry for the short hiatus on this. I cannot seem to get the Travis CI build to pass. It passes without a problem on my local machine. Aside from that I think the branch is ready to be merged right?

**Comment by amirci** [comment]
I'll look into the CI problem.... I thought we had a .travis.yml config file....

**Comment by gmcgibbon** [comment]
We do and I didn't change the .travis.yml file in any way. I had to reinstall ruby recently to enable shared library support for the Qt gem. However, this is a change made to my local machine only and should not break winnivote or its Travis CI builds. I am guessing the [error](https://travis-ci.org/winnipegrb/winnivote/builds/25576075) raises due to winnivote's ruby version preference.

**Pull request closed without merging**